docs/proposals/rescheduling-for-critical-pods.md

```diff
@@ -0,0 +1,116 @@
+<!-- BEGIN MUNGE: UNVERSIONED_WARNING -->
+
+<!-- BEGIN STRIP_FOR_RELEASE -->
+
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+<img src="http://kubernetes.io/img/warning.png" alt="WARNING"
+     width="25" height="25">
+
+<h2>PLEASE NOTE: This document applies to the HEAD of the source tree</h2>
+
+If you are using a released version of Kubernetes, you should
+refer to the docs that go with that version.
+
+Documentation for other releases can be found at
+[releases.k8s.io](http://releases.k8s.io).
+</strong>
+--
+
+<!-- END STRIP_FOR_RELEASE -->
+
+<!-- END MUNGE: UNVERSIONED_WARNING -->
+
+# Rescheduler: guaranteed scheduling of critical addons
+
+## Motivation
+
+In addition to Kubernetes core components like api-server, scheduler, controller-manager running on a master machine
+there is a bunch of addons which due to various reasons have to run on a regular cluster node, not the master.
+Some of them are critical to have fully functional cluster: Heapster, DNS, UI. Users can break their cluster
+by evicting a critical addon (either manually or as a side effect of an other operation like upgrade)
+which possibly can become pending (for example when the cluster is highly utilized).
+To avoid such situation we want to have a mechanism which guarantees that
+critical addons are scheduled assuming the cluster is big enough.
+This possibly may affect other pods (including production user’s applications).
+
+## Design
+
+Rescheduler will ensure that critical addons are always scheduled.
```

Comment:
@timothysc Are you talking about the email I sent to kubernetes-sig-scheduling? I said
> nobody has started working on the "full" rescheduler or priority/preemption scheme yet; however there will be a simple version of preemption to allow prioritized scheduling of critical cluster addon pods in 1.4; see #29023 for details (design proposal coming very very soon)

This is the design proposal. I was referring to. As to whether this should be called "the rescheduler" I can see arguments both ways... It might be better to call this the "critical pod rescheduler" since the design does not explicitly consider the other types of rescheduling we have discussed elsewhere (e.g. in https://github.com/kubernetes/kubernetes/blob/master/docs/proposals/rescheduling.md) but I don't think it's a big deal. This is also not intended as a general priority/preemption mechanism.